system/src/ble_listening_mode_handler.cpp

```diff
@@ -198,49 +203,125 @@ int BleListeningModeHandler::applyControlRequestConfigurations() {
-    CHECK(hal_ble_gap_set_scan_response_data(ctrlReqSrData_.get(), ctrlReqSrDataLen_, nullptr));
+    CHECK_FALSE(exited_, SYSTEM_ERROR_INVALID_STATE);
+    bool success = false;
+    SCOPE_GUARD ({
+        if (!success) {
+            exit();
+        }
+    });
+    CHECK(hal_ble_lock(nullptr));
+    CHECK(hal_ble_notify_exit_listening_mode(nullptr));
+    CHECK(hal_ble_gap_set_advertising_data(ctrlReqAdvData_.data(), ctrlReqAdvData_.size(), nullptr));
+    CHECK(hal_ble_gap_set_scan_response_data(ctrlReqSrData_.data(), ctrlReqSrData_.size(), nullptr));
+    CHECK(hal_ble_notify_enter_listening_mode(nullptr));
+    CHECK(hal_ble_unlock(nullptr));
+    success = true;
     return SYSTEM_ERROR_NONE;
 }
 
 int BleListeningModeHandler::enter() {
+    exited_ = false;
+
+    bool success = false;
+    SCOPE_GUARD ({
+        if (!success) {
+            exit();
+        }
+    });
+
+    // Do not allow other thread to modify the BLE configurations.
+    CHECK(hal_ble_lock(nullptr));
+    // Now the BLE configurations are non-modifiable.
+    CHECK(hal_ble_notify_enter_listening_mode(nullptr));
     CHECK(constructControlRequestAdvData());
     CHECK(cacheUserConfigurations());
+    // Restore the user's BLE configurations on exited.
+    restoreUserConfig_ = true;
     CHECK(applyControlRequestAdvData());
     CHECK(applyControlRequestConfigurations());
-
     CHECK(hal_ble_set_callback_on_adv_events(onBleAdvEvents, this, nullptr));
-
     if (!preAdvertising_ && !preConnected_) {
         // Start advertising if it is neither connected nor advertising.
         CHECK(hal_ble_gap_start_advertising(nullptr));
     }
+    CHECK(hal_ble_unlock(nullptr));
+
+    success = true;
     return SYSTEM_ERROR_NONE;
 }
 
 int BleListeningModeHandler::exit() {
-    CHECK(restoreUserConfigurations());
-    CHECK(hal_ble_cancel_callback_on_adv_events(onBleAdvEvents, this, nullptr));
+    if (exited_) {
+        return SYSTEM_ERROR_NONE;
+    }
+
+    // FIXME: BLE is still connected to finalize the mobile setup when this function is called. In that case,
+    // there is a chance that the mobile setup process is terminated by other thread or user application.
+
+    preAdvData_.clear();
+    preSrData_.clear();
+    ctrlReqAdvData_.clear();
+    ctrlReqSrData_.clear();
+
+    // It should not CHECK() in this function, in case that the BLE HAL status cannot be restored correctly.
+
+    // Do not allow other thread to modify the BLE configurations until this function exits.
+    hal_ble_lock(nullptr);
+    // Now the BLE configurations are modifiable.
+    hal_ble_notify_exit_listening_mode(nullptr);
+    if (restoreUserConfig_) {
+        restoreUserConfig_ = false;
+        if (restoreUserConfigurations() != SYSTEM_ERROR_NONE) {
+            LOG(ERROR, "Failed to restore user configuration.");
+        }
+    }
+    hal_ble_cancel_callback_on_adv_events(onBleAdvEvents, this, nullptr);
+    hal_ble_unlock(nullptr);
+
+    exited_ = true;
```

Comment:
Let's set if before unlocking the BLE lock.

Reply:
If we used `BleLock`, it is always set before unlocking the BLE lock.